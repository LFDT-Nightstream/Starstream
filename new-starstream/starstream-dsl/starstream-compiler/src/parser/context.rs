use chumsky::{
    error::Rich,
    extra::{self, SimpleState},
    input,
};

/// Starsteam Parser State
#[derive(Default)]
pub struct State {}

impl State {
    pub fn new() -> SimpleState<Self> {
        SimpleState(State {})
    }
}

pub type Extra<'a> = extra::Full<Rich<'a, char>, SimpleState<State>, ()>;
#[allow(dead_code)]
pub type MapExtra<'a, 'b> = input::MapExtra<'a, 'b, &'a str, Extra<'a>>;
