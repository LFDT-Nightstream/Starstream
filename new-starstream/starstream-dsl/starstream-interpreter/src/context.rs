use std::collections::BTreeMap;

use starstream_types::StarstreamProgram;

#[derive(Debug, Clone)]
pub enum Value {
    // Primitive values
    Number(i64),
    Boolean(bool),
    // Linear handle types
    Utxo(u64),
}

/// Instruction context: manages which instructions are left to process
#[derive(Debug, Clone)]
pub struct InstructionContext {
    pub program: StarstreamProgram,
    // no need to explicitly track the program counter
    // it's implicit from the Rust program counter
}

#[derive(Debug, Clone, Default)]
pub struct ScopeStack {
    pub variables: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, Default)]
pub struct LocalMemoryContext {
    pub scope_stack: Vec<ScopeStack>,
}

/// Persistent memory context: shared variables that are globally referable
#[derive(Debug, Clone, Default)]
pub struct PersistentMemoryContext {
    pub global_variables: BTreeMap<String, i64>,
    pub heap: BTreeMap<usize, i64>,
    pub next_heap_address: usize,
}

#[derive(Debug, Clone)]
pub enum Type {
    Struct(Vec<(String, Type)>),
    Enum(Vec<(String, Vec<(String, Type)>)>),
}

/// Type context: manages type definitions and their identities
#[derive(Debug, Clone, Default)]
pub struct TypeContext {
    // Type definition and hash identity
    pub type_definitions: BTreeMap<String, (Type, u64)>,
}

/// Resource context: manages references to externally-managed resources
#[derive(Debug, Clone, Default)]
pub struct ResourceContext {
    pub tokens: Vec<i64>,
    // UTXOs are not ambient context.
}


/// Execution environment containing all contexts as defined in the language specification
#[derive(Debug, Clone)]
pub struct ExecutionEnvironment {
    pub instruction_context: InstructionContext,
    pub local_memory_context: LocalMemoryContext,
    pub persistent_memory_context: PersistentMemoryContext,
    pub type_context: TypeContext,
    pub resource_context: ResourceContext,
}

impl ExecutionEnvironment {
    pub fn new(program: StarstreamProgram) -> Self {
        Self {
            instruction_context: InstructionContext { program },
            local_memory_context: LocalMemoryContext::default(),
            persistent_memory_context: PersistentMemoryContext::default(),
            type_context: TypeContext::default(),
            resource_context: ResourceContext::default(),
        }
    }
}

