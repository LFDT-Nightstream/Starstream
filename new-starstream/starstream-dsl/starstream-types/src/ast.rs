//! Abstract syntax tree for the pared-down Starstream DSL.

use chumsky::span::SimpleSpan;
use serde::Serialize;

/// Entire program: a sequence of statements.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Program {
    pub statements: Vec<Statement>,
}

/// Statements allowed by the current grammar.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Statement {
    VariableDeclaration {
        name: Identifier,
        value: Spanned<Expr>,
    },
    Assignment {
        target: Identifier,
        value: Spanned<Expr>,
    },
    If {
        condition: Spanned<Expr>,
        then_branch: Block,
        else_branch: Option<Block>,
    },
    While {
        condition: Spanned<Expr>,
        body: Block,
    },
    Block(Block),
    Expression(Spanned<Expr>),
}

/// `{ statement* }`
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Block {
    pub statements: Vec<Statement>,
}

/// Expressions with precedence encoded via parser, not the AST shape.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Identifier(Identifier),
    Unary {
        op: UnaryOp,
        expr: Box<Spanned<Expr>>,
    },
    Binary {
        op: BinaryOp,
        left: Box<Spanned<Expr>>,
        right: Box<Spanned<Expr>>,
    },
    Grouping(Box<Spanned<Expr>>),
}

#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Literal {
    Integer(i64),
    Boolean(bool),
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub enum UnaryOp {
    Negate,
    Not,
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub enum BinaryOp {
    Multiply,
    Divide,
    Remainder,
    Add,
    Subtract,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    And,
    Or,
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq, Hash)]
pub struct Identifier {
    pub name: String,
    #[serde(skip)]
    pub span: Option<SimpleSpan>,
}

impl Identifier {
    pub fn new(name: impl Into<String>, span: Option<SimpleSpan>) -> Self {
        Self {
            name: name.into(),
            span,
        }
    }
}

/// Helper for attaching source spans to AST nodes.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Spanned<T> {
    pub node: T,
    #[serde(skip)]
    pub span: SimpleSpan,
}

impl<T> Spanned<T> {
    pub fn new(node: T, span: SimpleSpan) -> Self {
        Self { node, span }
    }

    pub fn none(node: T) -> Spanned<T> {
        Spanned {
            node,
            span: SimpleSpan {
                start: 0,
                end: 0,
                context: (),
            },
        }
    }
}
