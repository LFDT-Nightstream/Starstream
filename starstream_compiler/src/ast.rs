//! AST types describing a Starstream source file.

#[derive(Clone, Debug, Default)]
pub struct StarstreamProgram {
    pub items: Vec<ProgramItem>,
}

#[derive(Clone, Debug)]
pub enum ProgramItem {
    Utxo(Utxo),
    Script(Script),
    Token(Token),
}

#[derive(Clone, Debug)]
pub struct Utxo {
    pub name: Identifier,
    pub items: Vec<UtxoItem>,
}

#[derive(Clone, Debug)]
pub enum UtxoItem {
    Abi(Abi),
    Main(Main),
    Impl(Impl),
    Storage(Storage),
}

#[derive(Clone, Debug)]
pub struct Main {
    pub type_sig: Option<OptionallyTypedBindings>,
    pub block: Block,
}

#[derive(Clone, Debug)]
pub struct Token {
    pub name: Identifier,
    pub items: Vec<TokenItem>,
}

#[derive(Clone, Debug)]
pub enum TokenItem {
    Bind(Bind),
    Unbind(Unbind),
    Abi(Abi),
    Mint(Mint),
}

#[derive(Clone, Debug)]
pub struct Bind(pub Block);

#[derive(Clone, Debug)]
pub struct Unbind(pub Block);

#[derive(Clone, Debug)]
pub struct Mint(pub Block);

#[derive(Clone, Debug)]
pub struct Impl {
    pub name: Identifier,
    pub definitions: Vec<FnDef>,
}

#[derive(Clone, Debug)]
pub struct Script {
    pub definitions: Vec<FnDef>,
}

#[derive(Clone, Debug)]
pub struct Storage {
    pub bindings: TypedBindings,
}

#[derive(Clone, Debug)]
pub struct Sig {
    pub name: Identifier,
    pub input_types: Vec<Type>,
    pub output_type: Option<Type>,
}

#[derive(Clone, Debug)]
pub struct FnSig(pub Sig);

#[derive(Clone, Debug)]
pub struct FnDef {
    pub name: Identifier,
    pub inputs: OptionallyTypedBindings,
    pub output: Option<Type>,
    pub body: Block,
}

#[derive(Clone, Debug)]
pub enum EffectSig {
    EffectSig(Sig),
    EventSig(Sig),
    ErrorSig(Sig),
}

#[derive(Clone, Debug)]
pub enum AbiElem {
    FnSig(FnSig),
    EffectSig(EffectSig),
}

#[derive(Clone, Debug)]
pub struct Abi {
    pub values: Vec<AbiElem>,
}

#[derive(Clone, Debug)]
pub struct Identifier(pub String);

#[derive(Clone, Debug)]
pub enum Type {
    BaseType(Identifier, Option<Vec<Type>>),
    Object(TypedBindings),
    FnType(TypedBindings, Option<Box<Type>>),
}

#[derive(Clone, Debug)]
pub enum Statement {
    BindVar {
        var: Identifier,
        value: Expr,
        mutable: bool,
    },
    Return(Option<Expr>),
    Resume(Option<Expr>),
    Assign(Identifier, Expr),
    With(Block, Vec<(Effect, Block)>),
    While(Expr, LoopBody),
    Loop(LoopBody),
}

#[derive(Clone, Debug)]
pub enum LoopBody {
    Statement(Box<Statement>),
    Block(Block),
    Expr(Expr),
}

#[derive(Clone, Debug)]
pub enum Expr {
    Equals(Box<Self>, Box<Self>),
    NotEquals(Box<Self>, Box<Self>),
    LessThan(Box<Self>, Box<Self>),
    Add(Box<Self>, Box<Self>),
    Or(Box<Self>, Box<Self>),
    Sub(Box<Self>, Box<Self>),
    Pow(Box<Self>, Box<Self>),
    Mul(Box<Self>, Box<Self>),
    And(Box<Self>, Box<Self>),
    Div(Box<Self>, Box<Self>),
    Neg(Box<Self>),
    PrimaryExpr(
        PrimaryExpr,
        Option<Arguments>,
        Vec<(Identifier, Option<Arguments>)>,
    ),
    BlockExpr(BlockExpr),
}

#[derive(Clone, Debug)]
pub enum BlockExpr {
    IfThenElse(Box<Expr>, Box<Block>, Option<Box<Block>>),
    Block(Block),
}

#[derive(Clone, Debug)]
pub enum PrimaryExpr {
    Null,
    Number(f64),
    Bool(bool),
    Ident(Identifier),
    ParExpr(Box<Expr>),
    Yield(Box<Expr>),
    Raise(Box<Expr>),
    Object(Type, Vec<(Identifier, Expr)>),
}

#[derive(Clone, Debug)]
pub enum ExprOrStatement {
    Expr(Expr),
    Statement(Statement),
}

#[derive(Clone, Debug)]
pub enum Block {
    Chain {
        head: Box<ExprOrStatement>,
        tail: Box<Block>,
    },
    Close {
        semicolon: bool,
    },
}

#[derive(Clone, Debug)]
pub struct Arguments {
    pub xs: Vec<Expr>,
}

#[derive(Clone, Debug)]
pub struct OptionallyTypedBindings {
    pub values: Vec<(Identifier, Option<Type>)>,
}

#[derive(Clone, Debug)]
pub struct TypedBindings {
    pub values: Vec<(Identifier, Type)>,
}

#[derive(Clone, Debug)]
pub struct Effect {
    pub ident: Identifier,
    pub type_sig: OptionallyTypedBindings,
}
