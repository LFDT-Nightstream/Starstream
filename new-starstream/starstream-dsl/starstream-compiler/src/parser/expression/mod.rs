use chumsky::{prelude::*, span::SimpleSpan};
use starstream_types::ast::{BinaryOp, Expr, Spanned};

use super::context::Extra;

mod additive;
mod comparison;
mod equality;
mod logical_and;
mod logical_or;
mod multiplicative;
mod primary;
mod unary;

pub fn parser<'a>() -> impl Parser<'a, &'a str, Spanned<Expr>, Extra<'a>> {
    recursive(|expression| {
        let primary = primary::parser(expression.clone()).boxed();
        let unary = unary::parser(primary).boxed();
        let multiplicative = multiplicative::parser(unary).boxed();
        let additive = additive::parser(multiplicative).boxed();
        let comparison = comparison::parser(additive).boxed();
        let equality = equality::parser(comparison).boxed();
        let logical_and = logical_and::parser(equality).boxed();
        logical_or::parser(logical_and).boxed()
    })
    .boxed()
}

pub(super) fn chain<'a>(
    lower: impl Parser<'a, &'a str, Spanned<Expr>, Extra<'a>> + Clone + 'a,
    operator: impl Parser<'a, &'a str, BinaryOp, Extra<'a>> + Clone + 'a,
) -> impl Parser<'a, &'a str, Spanned<Expr>, Extra<'a>> {
    let tail = operator.then(lower.clone());

    lower
        .clone()
        .foldl(tail.repeated(), |lhs, (op, rhs)| combine(lhs, op, rhs))
}

fn combine(lhs: Spanned<Expr>, op: BinaryOp, rhs: Spanned<Expr>) -> Spanned<Expr> {
    let span = span_union(lhs.span, rhs.span);
    Spanned::new(
        Expr::Binary {
            op,
            left: Box::new(lhs),
            right: Box::new(rhs),
        },
        span,
    )
}

fn span_union(a: SimpleSpan, b: SimpleSpan) -> SimpleSpan {
    let start = a.start.min(b.start);
    let end = a.end.max(b.end);
    SimpleSpan::new((), start..end)
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;

    macro_rules! assert_expression_snapshot {
        ($code:expr) => {{
            let parsed = parser()
                .parse(indoc! { $code })
                .into_result()
                .expect("expression should parse");

            insta::with_settings!({
                description => format!("Code:\n\n{}", indoc! { $code }),
                omit_expression => true,
                prepend_module_to_snapshot => true,
            }, {
                insta::assert_debug_snapshot!(parsed);
            });
        }};
    }

    #[test]
    fn integer_literal() {
        assert_expression_snapshot!("42");
    }

    #[test]
    fn arithmetic_precedence() {
        assert_expression_snapshot!("1 + 2 * 3 - 4");
    }

    #[test]
    fn logical_chaining() {
        assert_expression_snapshot!("(1 < 2) && false || true");
    }
}
