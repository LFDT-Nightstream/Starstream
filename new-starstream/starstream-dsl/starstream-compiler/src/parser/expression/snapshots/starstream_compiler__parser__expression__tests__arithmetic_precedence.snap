---
source: starstream-compiler/src/parser/expression/mod.rs
description: "Code:\n\n1 + 2 * 3 - 4"
---
Spanned {
    node: Binary {
        op: Subtract,
        left: Spanned {
            node: Binary {
                op: Add,
                left: Spanned {
                    node: Literal(
                        Integer(
                            1,
                        ),
                    ),
                    span: 0..1,
                },
                right: Spanned {
                    node: Binary {
                        op: Multiply,
                        left: Spanned {
                            node: Literal(
                                Integer(
                                    2,
                                ),
                            ),
                            span: 4..5,
                        },
                        right: Spanned {
                            node: Literal(
                                Integer(
                                    3,
                                ),
                            ),
                            span: 8..9,
                        },
                    },
                    span: 4..9,
                },
            },
            span: 0..9,
        },
        right: Spanned {
            node: Literal(
                Integer(
                    4,
                ),
            ),
            span: 12..13,
        },
    },
    span: 0..13,
}
